specs/phase1/fraud-proofs.md

```diff
@@ -29,42 +32,259 @@ This document describes the shard transition function and fraud proofs as part o
+    return post_state
+```
+
+```python
+def compute_shard_transition_digest(beacon_state: BeaconState,
+                                    shard_state: ShardState,
+                                    beacon_parent_root: Root,
+                                    shard_body_root: Root) -> Bytes32:
+    # TODO: use SSZ hash tree root
+    return hash(
+        hash_tree_root(shard_state) + beacon_parent_root + shard_body_root
+    )
+```
+
+```python
+def verify_shard_block_message(beacon_state: BeaconState,
+                               shard_state: ShardState,
+                               block: ShardBlock,
+                               slot: Slot,
+                               shard: Shard) -> bool:
+    assert block.shard_parent_root == shard_state.latest_block_root
+    assert block.beacon_parent_root == get_block_root_at_slot(beacon_state, slot)
+    assert block.slot == slot
+    assert block.proposer_index == get_shard_proposer_index(beacon_state, slot, shard)
+    assert 0 < len(block.body) <= MAX_SHARD_BLOCK_SIZE
+    return True
+```
+
+```python
+def verify_shard_block_signature(beacon_state: BeaconState,
+                                 signed_block: SignedShardBlock) -> bool:
+    proposer = beacon_state.validators[signed_block.message.proposer_index]
+    domain = get_domain(beacon_state, DOMAIN_SHARD_PROPOSAL, compute_epoch_at_slot(signed_block.message.slot))
+    signing_root = compute_signing_root(signed_block.message, domain)
+    return bls.Verify(proposer.pubkey, signing_root, signed_block.signature)
+```
+
+### Verifying the proof
+
 TODO. The intent is to have a single universal fraud proof type, which contains the following parts:
 
-1. An on-time attestation on some `shard` signing a `ShardTransition`
-2. An index `i` of a particular position to focus on
-3. The `ShardTransition` itself
-4. The full body of the block
+1. An on-time attestation `attestation` on some shard `shard` signing a `transition: ShardTransition`
+2. An index `offset_index` of a particular position to focus on
+3. The `transition: ShardTransition` itself
+4. The full body of the shard block `shard_block`
 5. A Merkle proof to the `shard_states` in the parent block the attestation is referencing
+6. The `subkey` to generate the custody bit
+
+Call the following function to verify the proof:
 
-The proof verifies that one of the two conditions is false:
+```python
+def is_valid_fraud_proof(beacon_state: BeaconState,
+                         attestation: Attestation,
+                         offset_index: uint64,
+                         transition: ShardTransition,
+                         block: ShardBlock,
+                         subkey: BLSPubkey,
+                         beacon_parent_block: BeaconBlock) -> bool:
+    # 1. Check if `custody_bits[offset_index][j] != generate_custody_bit(subkey, block_contents)` for any `j`.
+    shard = get_shard(beacon_state, attestation)
```

Comment:
I think we can move `shard = get_shard(beacon_state, attestation)` to L136 to be closer to its usage